### DISCUSSION TITLE
Type error on invoke action implemented using the actions param

### DISCUSSION THREAD

**Discussion by BrunoQuaresma** [statelyai/xstate#2364] [Q&A] [answered]
Hi, I'm facing an error when I try to use an invoke action with an external implementation but I don't know if that is possible. 

Here is the code sample: https://codesandbox.io/s/quirky-bhabha-sv5e7?file=/src/searchMachine.ts:1577-1604

**Comment by davidkpiano** [comment]
The `searchMachine.ts` file is empty.

**Reply by BrunoQuaresma** [reply]
My bad, should be there now.

**Comment by davidkpiano** [comment, marked as answer]
The tricky thing with parameterized implementations (actions, guards, etc.) is that you have to assume that they can be called with _any_ event, so you should adjust your typings:

```ts
    actions: {
      // ...

      setResults: assign({
        results: (_ctx, event) => (event as DoneInvokeEvent<any>).data
      })
    },
    guards: {
      searchValid: (_ctx, event) => {
        if (event.type === 'SEARCH') {
          return (event.tweetId && event.tweetId.length > 0) as boolean;
        } else { return false }
      }
    }

**Reply by BrunoQuaresma** [reply]
Not related question but, is there any preference on use a `model` or not? I've been thinking it is harder to type states than using the "regular" way. 🤔

**Reply by davidkpiano** [reply]
> Not related question but, is there any preference on use a `model` or not? I've been thinking it is harder to type states than using the "regular" way. 🤔

Use whichever way works best. `model.assign(...)` actions are best placed as _inline actions_.

**Reply by BrunoQuaresma** [reply]
And about the model usage at all?

**Reply by davidkpiano** [reply]
Yes, the purpose of the model is to make TypeScript typings (both now and in the future) better. Its usage is not required.
